torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,803 @@
+            for p in self.params:
+                # e.g., when world_size == 1
+                if not p._is_sharded:  # type: ignore[attr-defined]
+                    continue
+                else:
+                    # Move params in CPU to CUDA for the all-gather.
+                    p_data = p.data.to(p._full_param_padded.device, non_blocking=True)  # type: ignore[attr-defined]
+                    p_full_size = p._full_param_padded.size()  # type: ignore[attr-defined]
+                    assert p_full_size.numel() == p_data.numel() * self.world_size
+                    assert p._full_param_padded.storage().size() == 0  # type: ignore[attr-defined]
+                    # Allocate based on full size from all shards.
+                    _alloc_storage(p._full_param_padded, size=p_full_size)  # type: ignore[attr-defined]
+                    output_tensor = p._full_param_padded  # type: ignore[attr-defined]
+
+                    # Fill output_tensor with (p.data for each shard in self.world_size)
+                    dist._all_gather_base(output_tensor, p_data, group=self.process_group)
+
+                    # Set p.data = output_tensor (with padding trimmed)
+                    update_p_data(output_tensor)
+
+        torch.cuda.current_stream().wait_stream(self._streams["all_gather"])
+
+    @torch.no_grad()
+    def _use_full_params(self) -> None:
+        """
+        Switch p.data pointers to use the full params.
+        Note: this assumes full params are already gathered.
+
+        Note: this might be called after full_params is already in used. So please
+              make sure it is idempotent in that case.
+        """
+        for p in self.params:
+            assert p._full_param_padded.storage().size() != 0  # type: ignore[attr-defined]
+            p.data = p._full_param_padded[: p._orig_size.numel()].view(p._orig_size)  # type: ignore[attr-defined]
+
+    @torch.no_grad()
+    def _prep_grads_for_backward(self) -> None:
+        """Make sure p.grad has the correct size/device, otherwise set it to None."""
+        for p in self.params:
+            if p.grad is not None and (
+                p.grad.size() != p._orig_size  # type: ignore[attr-defined]
+                or p.grad.device != p.device
+            ):
+                p.grad = None
+
+    @torch.no_grad()
+    def _free_full_params(self, params: Optional[List[Parameter]] = None) -> None:
+        """Free up storage for full parameters."""
+        if params is None:
+            params = self.params
+        current_stream = torch.cuda.current_stream()
+        for p in params:
+            # e.g., world_size == 1
+            if not p._is_sharded:  # type: ignore[attr-defined]
+                continue
+            # Don't let PyTorch reuse this memory until all work in the current
+            # stream is complete.
+            p._full_param_padded.record_stream(current_stream)  # type: ignore[attr-defined]
```

Comment:
`_full_param_padded` is also used in `allgather` stream. I assume we don't need to record `allgather` stream because there is always computation on this tensor on the current stream after allgather? What if the params of an FSDP united is unused in forward? Will this free pending tensors for allgather?

Reply:
that is correct, we do not need to record allgather, as it is blocking and computation depends on it.  

that is really good catch for free full param when params are not used. In this case, full params are not freed if params are not used.  However, once we flatten these params later on, the full flat_param will be freed unless all the params are not used